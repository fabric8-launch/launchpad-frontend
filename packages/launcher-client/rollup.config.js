import peerDepsExternal from 'rollup-plugin-peer-deps-external';
import filesize from 'rollup-plugin-filesize';
import tslint from 'rollup-plugin-tslint';
import typescript from 'rollup-plugin-typescript2';
import json from 'rollup-plugin-json';
import stripCode from "rollup-plugin-strip-code"

import pkg from './package.json';
const isProduction = process.env.NODE_ENV === 'production';

const config = {
    input: 'src/index.ts',
    output: [
        {
            file: pkg.main,
            format: 'cjs',
        },
        {
            file: pkg.module,
            format: 'es',
        },
    ],
    external: [
        ...Object.keys(pkg.dependencies || {}),
        ...Object.keys(pkg.peerDependencies || {}),
    ],
    plugins: [
        peerDepsExternal(),
        tslint({
            include: './src/**/*.tsx?'
        }),
        typescript({
            clean: true,
            typescript: require('typescript'),
            verbosity: 0,
            useTsconfigDeclarationDir: true
        }),
        json(),
        isProduction && stripCode({
            start_comment: 'test-code',
            end_comment: 'end-test-code'
        }),
        filesize(),
    ],
};

export default config;
