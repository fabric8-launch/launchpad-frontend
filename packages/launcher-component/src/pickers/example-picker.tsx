import * as React from 'react';
import {
  DataList,
  DataListCell,
  DataListContent,
  DataListItem,
  Radio,
  Title,
  FormSelect,
  FormSelectOption,
  Button,
  Split,
  SplitItem
} from '@patternfly/react-core';
import { Catalog, filter, ExampleRuntime, ExampleMission } from 'launcher-client';
import { InputProps, Picker } from '../core/types';
import _ from 'lodash';

export interface ExamplePickerValue {
  missions?: ExampleMission[];
  missionId?: string;
  runtimeId?: string;
  versionId?: string;
}

interface ExamplePickerProps extends InputProps<ExamplePickerValue> {
  catalog: Catalog;
}

export const ExamplePicker: Picker<ExamplePickerProps, ExamplePickerValue> = {
  checkCompletion: value => !!value.missionId && !!value.runtimeId && !!value.versionId,
  Element: props => {
    const query = { runtime: { id: '', name: '', version: { id: '', name: '' } } };
    const missionQuery = { mission: { id: '', name, description: '', runtime: { id: '', name, version: { name, id: '' } } } };
    const runtimesMap = _.keyBy(filter(query, props.catalog) as ExampleRuntime[], 'id');
    const missions = !props.value.missions ? filter(missionQuery, props.catalog) as ExampleMission[] : props.value.missions;
    const filterCatalog = (runtime) => {
      props.onChange({
        runtimeId: runtime,
        missions: (filter({
          mission: {
            id: '', name, description: '', runtime:
              { id: runtime, name, version: { name, id: '' } }
          }
        }, props.catalog) as ExampleMission[]).filter(m => !!m.runtime)
      });
    };

    return (
      <React.Fragment>
        <Split>
          <SplitItem isMain>
            Already decided what runtime you want to use? Select it here:
        </SplitItem>
          <SplitItem isMain={false}>
            <Button variant="tertiary" onClick={() => props.onChange({ ...props.value, missions: undefined, runtimeId: '' })}>Any</Button>
            {_.map(runtimesMap).map((runtime, index) => (
              <Button
                style={{marginLeft: '5px'}}
                variant="tertiary"
                key={index}
                onClick={() => filterCatalog(runtime.id)}
                isDisabled={props.value.runtimeId === runtime.id && !!props.value.missions}
              >
                {runtime.name}
              </Button>
            ))}
          </SplitItem>
        </Split>
        <DataList aria-label="select-mission" style={{marginTop: '20px'}}>
          {
            missions!.map((mission, i) => {
              const isSelected = props.value.missionId === mission.id;
              const onChangeSelected = () => {
                props.onChange({ ...props.value, missionId: mission.id });
              };
              return (
                <DataListItem
                  isExpanded={isSelected}
                  aria-labelledby={mission.name}
                  value={mission.id}
                  key={i}
                  style={{ cursor: 'pointer' }}
                >
                  <DataListCell width={1} style={{ flex: 'none' }}>
                    <Radio
                      aria-label={`Choose ${mission.name} as mission`}
                      value={mission.id}
                      checked={isSelected}
                      onChange={onChangeSelected}
                      name="mission"
                      id={`radio-choose-${mission.id}-as-mission`}
                    />
                  </DataListCell>
                  <DataListCell width={1} onClick={onChangeSelected}><Title size="lg">{mission.name}</Title></DataListCell>
                  <DataListCell width={2} onClick={onChangeSelected}>{mission.description}</DataListCell>
                  <DataListContent aria-label={'Detail for ' + mission.name} isHidden={!isSelected}>
                    <FormSelect
                      isDisabled={mission.runtime!.length === 1}
                      id={mission.id + 'runtime-select'}
                      value={props.value.runtimeId}
                      onChange={value => props.onChange({ ...props.value, runtimeId: value })}
                      aria-label="Select Runtime"
                    >
                      {mission.runtime!.length !== 1 && <FormSelectOption
                        value=""
                        label="Select a Runtime"
                      />}
                      {mission.runtime!.map((runtime, index) => (
                        <FormSelectOption
                          key={index}
                          value={runtime.id}
                          label={runtime.name}
                        />
                      ))
                      }
                    </FormSelect>
                    {props.value.runtimeId &&
                      <FormSelect
                        id={mission.id + 'version-select'}
                        value={props.value.versionId}
                        onChange={value => props.onChange({ ...props.value, versionId: value })}
                        aria-label="Select Version"
                      >
                        <FormSelectOption
                          key={-1}
                          value=""
                          label="Select a Version"
                        />
                        {(runtimesMap[props.value.runtimeId] as any).version.map((version, index) => (
                          <FormSelectOption
                            key={index}
                            value={version.id}
                            label={version.name}
                          />
                        ))
                        }
                      </FormSelect>
                    }
                  </DataListContent>
                </DataListItem>
              );
            })
          }
        </DataList>
      </React.Fragment>
    );
  }

};
