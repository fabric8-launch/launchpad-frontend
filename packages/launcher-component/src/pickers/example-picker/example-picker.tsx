import * as React from 'react';
import {
  DataList,
  DataListCell,
  DataListContent,
  DataListItem,
  Radio,
  Title,
  FormSelect,
  FormSelectOption
} from '@patternfly/react-core';
import { Catalog, filter, ExampleRuntime } from 'launcher-client';
import { InputProps } from '../../core/types';
import _ from 'lodash';

export interface ExamplePickerValue {
  missionId?: string;
  runtimeId?: string;
  versionId?: string;
}

interface ExamplePickerProps extends InputProps<ExamplePickerValue> {
  catalog: Catalog;
}

export function ExamplePicker(props: ExamplePickerProps) {
  const query = { runtime: { id: '', name: '', version: {id: '', name: ''} } };
  const runtimesMap = _.keyBy(filter(query, props.catalog) as ExampleRuntime[], 'id');

  return (
    <React.Fragment>
      <DataList aria-label="select-mission">
        {
          props.catalog.missions.map((mission, i) => {
            const isSelected = props.value.missionId === mission.id;
            const onChangeSelected = () => {
              props.onChange({ missionId: mission.id });
            };
            return (
              <DataListItem
                isExpanded={isSelected}
                aria-labelledby={mission.name}
                value={mission.id}
                key={i}
                style={{ cursor: 'pointer' }}
              >
                <DataListCell width={1} style={{ flex: 'none' }}>
                  <Radio
                    aria-label={`Choose ${mission.name} as mission`}
                    value={mission.id}
                    checked={isSelected}
                    onChange={onChangeSelected}
                    name="mission"
                    id={`radio-choose-${mission.id}-as-mission`}
                  />
                </DataListCell>
                <DataListCell width={1} onClick={onChangeSelected}><Title size="lg">{mission.name}</Title></DataListCell>
                <DataListCell width={2} onClick={onChangeSelected}>{mission.description}</DataListCell>
                <DataListContent aria-label={'Detail for ' + mission.name} isHidden={!isSelected}>
                  <FormSelect
                    id={mission.id + 'runtime-select'}
                    value={props.value.runtimeId}
                    onChange={value => props.onChange({ ...props.value, runtimeId: value })}
                    aria-label="Select Runtime"
                  >
                    {_.map(runtimesMap).map((runtime, index) => {
                      if (!props.value.runtimeId) {
                        props.value.runtimeId = runtime.id;
                      }
                      return (
                        <FormSelectOption
                          key={index}
                          value={runtime.id}
                          label={runtime.name}
                        />
                      );
                    }
                    )}
                  </FormSelect>
                  {props.value.runtimeId &&
                    <FormSelect
                      id={mission.id + 'version-select'}
                      value={props.value.versionId}
                      onChange={value => props.onChange({ ...props.value, versionId: value })}
                      aria-label="Select Version"
                    >
                      {(runtimesMap[props.value.runtimeId] as any).version.map((version, index) => {
                        if (!props.value.versionId) {
                          props.value.versionId = version.id;
                        }
                        return (
                          <FormSelectOption
                            key={index}
                            value={version.id}
                            label={version.name}
                          />
                        );
                      }
                      )}
                    </FormSelect>
                  }
                </DataListContent>
              </DataListItem>
            );
          })
        }
      </DataList>
    </React.Fragment>
  );
}
