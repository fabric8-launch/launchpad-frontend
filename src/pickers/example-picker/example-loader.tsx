import React from 'react';
import * as _ from 'lodash';
import { Example, ExampleMission } from 'launcher-client';
import { useLauncherClient } from '../../launcher-client-context';
import { DataLoader } from '../../core/data-loader/data-loader';

export function createMissions(examples: Example[]): ExampleMission[] {
  const groupedByMission = _.groupBy(examples, (e) => e.mission.id);
  return _.values(groupedByMission).map((missionBoosters) => {
    const mission = _.first(missionBoosters).mission;
    return {
      id: mission.id,
      name: mission.name,
      description: mission.description,
      community: false,
      advanced: _.get(mission, 'metadata.level') === 'advanced' || _.get(mission, 'metadata.prerequisite', false),
      suggested: _.get(mission, 'metadata.suggested', false),
      showMore: false,
      disabled: true,
      boosters: missionBoosters
    };
  });
}

export function ExamplesLoader(props: {children: (missions: ExampleMission[]) => any }) {
  const client = useLauncherClient();
  const itemsLoader = () => client.exampleCatalog().then(examples => {
    return createMissions(examples);
  });
  return (
    <DataLoader loader={itemsLoader} default={[]} >
      {props.children}
    </DataLoader>
  );
}
