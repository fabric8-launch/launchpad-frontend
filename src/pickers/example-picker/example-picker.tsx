import * as React from "react";
import { DataList, DataListItem, DataListCell, Title, DataListCheck, DataListToggle, DataListContent } from "@patternfly/react-core";
import { Example, ExampleMission } from "launcher-client";
import { InputProps } from "../../core/types";

export type ExamplePickerValue = Example;

interface ExamplePickerProps extends InputProps<ExamplePickerValue> {
  examples: Example[];
  missions: ExampleMission[];
}

export function ExamplePicker(props: ExamplePickerProps) {
  const onChangeSelected = (selected) => {
    props.onChange({ ...props.value, selected })
  };
  return (
    <React.Fragment>
      <DataList aria-label="select-mission">
        {
          props.missions.map((mission, i) => (
            <DataListItem isExpanded={props.value.name === mission.name} aria-labelledby={mission.name} value={mission.id}
              {...mission}
              key={i}>
              <DataListToggle
                isExpanded={props.value.name === mission.name}
                id={mission.id}
                aria-labelledby={mission.id}
                aria-label={"Toggle details for " + mission.name}
              />
              <DataListCheck aria-labelledby={mission.name} name={mission.id} onChange={onChangeSelected}/>
              <DataListCell width={1}><Title size="lg">{mission.name}</Title></DataListCell>
              <DataListCell width={2}>{mission.description}</DataListCell>
              <DataListContent aria-label={"Detail for " + mission.name} isHidden={props.value.name !== mission.name}>
                <p>
                  Version and runtime selected...
                </p>
              </DataListContent>
            </DataListItem>
          ))
        }
      </DataList>
    </React.Fragment>
  );
}