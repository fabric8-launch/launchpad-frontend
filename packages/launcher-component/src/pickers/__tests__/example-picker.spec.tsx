import * as React from 'react';
import { cleanup, render, fireEvent } from 'react-testing-library';
import { ExamplePicker } from '../example-picker';
import { ExampleMission } from 'launcher-client';
import { FormPanel } from '../../core/form-panel/form-panel';

afterEach(cleanup);

describe('<ExamplePicker />', () => {
  it('renders the ExamplePicker correctly', () => {
    const comp = render(<ExamplePicker.Element missions={missions} value={{}} onChange={() => {}}/>);
    expect(comp.asFragment()).toMatchSnapshot();
  });

  it('show the version dropdown', () => {
    const handleSave = jest.fn();
    const comp = render(
      <FormPanel
        initialValue={{}}
        validator={ExamplePicker.checkCompletion}
        onSave={handleSave}
        onCancel={() => {}}
      >
        {(inputProps) => (<ExamplePicker.Element missions={missions} {...inputProps}/>)}
      </FormPanel>
    );
    const missionRadio = comp.getByLabelText('Choose Mission Name as mission');
    fireEvent.click(missionRadio);
    expect(comp.getByLabelText('Select Runtime')).toBeDefined();
    fireEvent.change(comp.getByLabelText('Select Runtime'), { target: { value: 'nodejs' } });
    expect(comp.getByDisplayValue('10.x (Community)')).toBeDefined();
    fireEvent.change(comp.getByLabelText('Select Version'), { target: { value: 'community' } });
    fireEvent.click(comp.getByText('Save'));
    expect(handleSave).toHaveBeenCalledTimes(1);
    expect(comp.asFragment()).toMatchSnapshot();
  });

});

const missions: ExampleMission[] = [
  {id: 'test', description: 'bla', name: 'Mission Name', metadata: {}, runtime: [
    {id: 'nodejs', name: 'Node', icon: '', versions: [
      {id: 'community', name: '10.x (Community)'}
    ]}
  ]}
];
