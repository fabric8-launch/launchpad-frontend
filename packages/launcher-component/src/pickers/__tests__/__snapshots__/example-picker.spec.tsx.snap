// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExamplePicker /> renders the ExamplePicker correctly 1`] = `
<DocumentFragment>
  <div
    class="pf-l-split"
  >
    <div
      class="pf-l-split__item pf-m-main"
    >
      Already decided what runtime you want to use? Select it here:
    </div>
    <div
      class="pf-l-split__item"
    >
      <button
        class="pf-c-button pf-m-tertiary"
        type="button"
      >
        Any
      </button>
      <button
        class="pf-c-button pf-m-tertiary"
        style="margin-left: 5px;"
        type="button"
      >
        Node
      </button>
    </div>
  </div>
  <ul
    aria-label="select-mission"
    class="pf-c-data-list"
    role="list"
    style="margin-top: 20px;"
  >
    <li
      aria-labelledby="Mission Name"
      class="pf-c-data-list__item"
      style="cursor: pointer;"
      value="test"
    >
      <div
        class="pf-c-data-list__cell"
        style="flex: 0 0 auto;"
      >
        <div
          class="pf-c-radio"
        >
          <input
            aria-invalid="false"
            aria-label="Choose Mission Name as mission"
            class="pf-c-radio__input"
            id="radio-choose-test-as-mission"
            name="mission"
            type="radio"
            value="test"
          />
        </div>
      </div>
      <div
        class="pf-c-data-list__cell"
      >
        <h1
          class="pf-c-title pf-m-lg"
        >
          Mission Name
        </h1>
      </div>
      <div
        class="pf-c-data-list__cell pf-m-flex-2"
      >
        bla
      </div>
      <section
        aria-label="Detail for Mission Name"
        class="pf-c-data-list__expandable-content"
        hidden=""
        id=""
      >
        <select
          aria-invalid="false"
          aria-label="Select Runtime"
          class="pf-c-form-control"
          disabled=""
          id="testruntime-select"
        >
          <option
            class=""
            value="nodejs"
          >
            Node
          </option>
        </select>
      </section>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`<ExamplePicker /> show the version dropdown 1`] = `
<DocumentFragment>
  <div
    class="form-panel"
    style="padding: 20px;"
  >
    <div
      class="pf-l-split"
    >
      <div
        class="pf-l-split__item pf-m-main"
      >
        Already decided what runtime you want to use? Select it here:
      </div>
      <div
        class="pf-l-split__item"
      >
        <button
          class="pf-c-button pf-m-tertiary"
          type="button"
        >
          Any
        </button>
        <button
          class="pf-c-button pf-m-tertiary"
          style="margin-left: 5px;"
          type="button"
        >
          Node
        </button>
      </div>
    </div>
    <ul
      aria-label="select-mission"
      class="pf-c-data-list"
      role="list"
      style="margin-top: 20px;"
    >
      <li
        aria-labelledby="Mission Name"
        class="pf-c-data-list__item pf-m-expanded"
        style="cursor: pointer;"
        value="test"
      >
        <div
          class="pf-c-data-list__cell"
          style="flex: 0 0 auto;"
        >
          <div
            class="pf-c-radio"
          >
            <input
              aria-invalid="false"
              aria-label="Choose Mission Name as mission"
              class="pf-c-radio__input"
              id="radio-choose-test-as-mission"
              name="mission"
              type="radio"
              value="test"
            />
          </div>
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <h1
            class="pf-c-title pf-m-lg"
          >
            Mission Name
          </h1>
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          bla
        </div>
        <section
          aria-label="Detail for Mission Name"
          class="pf-c-data-list__expandable-content"
          id=""
        >
          <select
            aria-invalid="false"
            aria-label="Select Runtime"
            class="pf-c-form-control"
            disabled=""
            id="testruntime-select"
          >
            <option
              class=""
              value="nodejs"
            >
              Node
            </option>
          </select>
          <select
            aria-invalid="false"
            aria-label="Select Version"
            class="pf-c-form-control"
            id="testversion-select"
          >
            <option
              class=""
              value="community"
            >
              10.x (Community)
            </option>
          </select>
        </section>
      </li>
    </ul>
    <hr
      class="separator"
    />
    <div
      class="pf-l-toolbar"
    >
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Save form"
          class="pf-c-button pf-m-primary"
          type="button"
        >
          Save
        </button>
      </div>
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Cancel form"
          class="pf-c-button pf-m-secondary"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
